### PR TITLE
Simple API Doc generation

### PR DESCRIPTION
**Pull request by robfletcher**
This PR creates an endpoint that returns the JSON schema for `SubmittedDeliveryConfig`. The springdoc-openapi solution in place previously required a lot of custom handling and ended up documenting the entire REST API when all we really need is the file format for delivery configs.

There's no doubt still some tuning to do here, but it's a good start and should give us something usable. Still outstanding:

- [x] Support custom deserialization of `TitusClusterSpec`

The schema generated currently can be seen [here](https://gist.github.com/robfletcher/0407d24577707ba5770c3c7b3e43aaf6)

### CODE REVIEW

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/Constraint.kt [hunk 1]
```diff
@@ -1,4 +1,6 @@
 package com.netflix.spinnaker.keel.api
 
-abstract class Constraint(val type: String)
+import com.netflix.spinnaker.keel.api.schema.Discriminator
+
+abstract class Constraint(@Discriminator val type: String)
```
Comment (robfletcher) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/Constraint.kt:5]: Any class that has sub-types that are determined by a discriminator property needs to have this annotation or the JSON schema generation will blow up.

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/schema/Factory.kt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package com.netflix.spinnaker.keel.api.schema
+
+import kotlin.annotation.AnnotationTarget.CONSTRUCTOR
+
+@Target(CONSTRUCTOR)
+annotation class Factory
```
Comment (robfletcher) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/schema/Factory.kt:6]: Effectively the same thing as `@JsonCreator` without needing to bring in `jackson-annotations`

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/KeelApiModule.kt [hunk 3]
```diff
@@ -63,6 +70,14 @@ object KeelApiModule : SimpleModule("Keel API") {
...
       setMixInAnnotations<Resource<*>, ResourceMixin>()
       setMixInAnnotations<ResourceSpec, ResourceSpecMixin>()
+      insertAnnotationIntrospector(object : NopAnnotationIntrospector() {
+        override fun findCreatorAnnotation(config: MapperConfig<*>, a: Annotated): JsonCreator.Mode? =
+          when {
+            a.hasAnnotation(Factory::class.java) -> PROPERTIES
+            a.rawType.constructors.any { it.isAnnotationPresent(Factory::class.java) } -> DISABLED
+            else -> super.findCreatorAnnotation(config, a)
+          }
+      })
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/KeelApiModule.kt:80]: This will mean we treat `@Factory` the same as `@JsonCreator`

File: keel-ec2-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ClusterSpec.kt [hunk 4]
```diff
@@ -124,6 +126,35 @@ data class ClusterSpec(
...
+    ServerGroupSpec(
+      launchConfiguration,
+      capacity,
+      dependencies,
+      health,
+      scaling,
+      tags
+    ),
+    overrides
+  )
```
Comment (robfletcher) [inline keel-ec2-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ClusterSpec.kt:156]: Secondary constructor for `ClusterSpec` without needing a custom deserializer or Jackson annotation

File: keel-schema-generator/src/main/kotlin/com/netflix/spinnaker/keel/schema/ResourceKindSchemaCustomizer.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  private val kinds: Iterable<ResourceKind>
+) : SchemaCustomizer {
+
+  override fun supports(type: KClass<*>) = type == ResourceKind::class
+
+  override fun buildSchema() = EnumSchema(
+    description = "The resource kind that corresponds with each spec format.",
+    enum = kinds.map(ResourceKind::toString)
+  )
+}
```
Comment (robfletcher) [inline keel-schema-generator/src/main/kotlin/com/netflix/spinnaker/keel/schema/ResourceKindSchemaCustomizer.kt:17]: This is the only bit of customization necessary. I could even add some kind of annotation for this but this doesn't seem too heinous.

File: keel-schema-generator/src/main/kotlin/com/netflix/spinnaker/keel/schema/Schema.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
+package com.netflix.spinnaker.keel.schema
+
+import com.fasterxml.jackson.annotation.JsonValue
+import java.util.SortedMap
+import java.util.SortedSet
+
+interface Schema {
+  val description: String?
+}
```
Comment (robfletcher) [inline keel-schema-generator/src/main/kotlin/com/netflix/spinnaker/keel/schema/Schema.kt:9]: This file contains all the model classes I use for generating the JSON schema

File: keel-web/keel-web.gradle.kts [hunk 7]
```diff
@@ -28,8 +29,7 @@ dependencies {
   implementation("com.netflix.spinnaker.fiat:fiat-api:${property("fiatVersion")}")
   implementation("com.netflix.spinnaker.fiat:fiat-core:${property("fiatVersion")}")
   implementation("net.logstash.logback:logstash-logback-encoder")
-  implementation("org.springdoc:springdoc-openapi-webmvc-core")
-  implementation("org.springdoc:springdoc-openapi-kotlin")
+  implementation("io.swagger.core.v3:swagger-annotations:2.1.2")
```
Comment (robfletcher) [inline keel-web/keel-web.gradle.kts:32]: No more Springdoc OpenAPI but since we use some Swagger annotations on our controller endpoints I added that dependency explicitly (we no longer get it transitively)

File: keel-web/keel-web.gradle.kts [hunk 8]
```diff
@@ -51,7 +51,7 @@ dependencies {
   testImplementation("com.netflix.spinnaker.kork:kork-security")
   testImplementation("com.squareup.okhttp3:mockwebserver")
   testImplementation("org.testcontainers:mysql")
-  testImplementation("org.openapi4j:openapi-schema-validator:0.7")
+  testImplementation("com.networknt:json-schema-validator:1.0.43")
```
Comment (robfletcher) [inline keel-web/keel-web.gradle.kts:54]: Validator lib for JSON schema rather than Open API. This is used in `ApiDocCompatibilityTests`

File: keel-web/src/main/kotlin/com/netflix/spinnaker/config/DefaultConfiguration.kt [hunk 9]
```diff
@@ -96,4 +101,18 @@ class DefaultConfiguration(
...
+    resourceHandlers: List<ResourceHandler<*, *>>,
+    migrators: List<SpecMigrator<*, *>>
+  ): Generator = Generator(
+    extensionRegistry = extensionRegistry,
+    schemaCustomizers = listOf(
+      ResourceKindSchemaCustomizer(
+        resourceHandlers.map { it.supportedKind.kind } + migrators.map { it.input.kind }
+      )
+    )
+  )
```
Comment (robfletcher) [inline keel-web/src/main/kotlin/com/netflix/spinnaker/config/DefaultConfiguration.kt:117]: This is the schema generator bean

File: keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/DeliveryConfigController.kt [hunk 10]
```diff
@@ -138,4 +142,13 @@ class DeliveryConfigController(
...
   }
+
+  @Operation(
+    description = "Responds with the JSON schema for POSTing to /delivery-configs"
+  )
+  @GetMapping(
+    path = ["/schema"],
+    produces = [APPLICATION_JSON_VALUE, APPLICATION_YAML_VALUE]
+  )
+  fun schema(): RootSchema = generator.generateSchema<SubmittedDeliveryConfig>()
```
Comment (robfletcher) [inline keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/DeliveryConfigController.kt:153]: The endpoint where the schema is output

File: keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocCompatibilityTests.kt [hunk 11]
```diff
@@ -33,90 +29,55 @@ import strikt.api.expectThat
     "keel.plugins.ec2.enabled=true",
     "keel.plugins.titus.enabled=true"
   ],
-  webEnvironment = MOCK
+  webEnvironment = NONE
 )
 @AutoConfigureMockMvc
 @EnableAutoConfiguration(exclude = [TaskSchedulingAutoConfiguration::class])
-class ApiDocCompatibilityTests : JUnit5Minutests {
+class ApiDocCompatibilityTests {
```
Comment (robfletcher) [inline keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocCompatibilityTests.kt:36]: This test is pretty valuable as it ensures the schema we generate will validate a delivery config. We can extend this with more edge cases ideally.

File: keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocTests.kt [hunk 12]
```diff
@@ -64,142 +98,131 @@ import strikt.jackson.textValues
...
-  webEnvironment = MOCK
+  webEnvironment = NONE
 )
 @AutoConfigureMockMvc
 @EnableAutoConfiguration(exclude = [TaskSchedulingAutoConfiguration::class])
 class ApiDocTests : JUnit5Minutests {
+
   @Autowired
-  lateinit var mvc: MockMvc
+  lateinit var generator: Generator
```
Comment (robfletcher) [inline keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocTests.kt:108]: This set of tests are sanity checks that the JSON schema renders how we expect. Fundamentally it's testing the same things as before. It no longer needs to hit a controller endpoint to get the schema as we have a generator bean. Also some of the schema structure has changed a bit due to the differences between Open API and JSON schema

**Comment by gal-yardeni** [comment]
This is amazing 😄 It feels a bit like something that should live outside of keel -- are other services will be able to utilize this?

**Comment by robfletcher** [comment]
> This is amazing 😄 It feels a bit like something that should live outside of keel -- are other services will be able to utilize this?

Thanks. I feel like there is potential for making it a library. It is _very_ dependent on Kotlin though. It's also somewhat tied to the fact we're only really interested in the one input format. A few assumptions / limitations:

1. I assume that "properties" means "constructor parameters" because this is very much only intended for creating a schema for inbound delivery configs. If you wanted to use this for documenting schemas for outbound stuff that assumption wouldn't necessarily make sense. For most things it would be the same but there are exceptions such as `ClusterSpec` that uses a secondary constructor because the model doesn't quite match the JSON format.
2. Outside of Kotlin you can't get the name of a constructor parameter so that approach wouldn't work.
3. Default values and nullability let me determine if a property is optional, again that wouldn't work outside of Kotlin.
4. I totally rely on Keel's `ExtensionRegistry` (which could also live outside of Keel but has its own assumptions about how we want extensibility to work).

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/schema/Factory.kt (hunk 2 above)
Comment (luispollo) [reply keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/schema/Factory.kt:6]: I was debating with myself whether this should be named something more specific, given its purpose... (like `JsonFactory` or `JsonConstructor` perhaps). Probably not worth the energy, but thought I'd mention it.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/KeelApiModule.kt [hunk 13]
```diff
@@ -63,6 +70,14 @@ object KeelApiModule : SimpleModule("Keel API") {
       setMixInAnnotations<SubnetAwareRegionSpec, SubnetAwareRegionSpecMixin>()
       setMixInAnnotations<Resource<*>, ResourceMixin>()
       setMixInAnnotations<ResourceSpec, ResourceSpecMixin>()
+      insertAnnotationIntrospector(object : NopAnnotationIntrospector() {
+        override fun findCreatorAnnotation(config: MapperConfig<*>, a: Annotated): JsonCreator.Mode? =
+          when {
+            a.hasAnnotation(Factory::class.java) -> PROPERTIES
+            a.rawType.constructors.any { it.isAnnotationPresent(Factory::class.java) } -> DISABLED
```
Comment (luispollo) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/KeelApiModule.kt:77]: Do you mind explaining what these two cases account for?

File: keel-ec2-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ClusterSpec.kt [hunk 14]
```diff
@@ -106,9 +108,9 @@ private fun ClusterSpec.resolveHealth(region: String): Health {
     cooldown = overrides[region]?.health?.cooldown ?: defaults.health?.cooldown ?: default.cooldown,
     warmup = overrides[region]?.health?.warmup ?: defaults.health?.warmup ?: default.warmup,
     healthCheckType = overrides[region]?.health?.healthCheckType ?: defaults.health?.healthCheckType
-      ?: default.healthCheckType,
+    ?: default.healthCheckType,
```
Comment (luispollo) [inline keel-ec2-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ClusterSpec.kt:111]: nit: Original indentation looked right...

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/schema/Factory.kt (hunk 2 above)
Comment (robfletcher) [reply keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/schema/Factory.kt:6]: I didn't want anyone to mistake it for an actual Jackson annotation

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/KeelApiModule.kt (hunk 13 above)
Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/KeelApiModule.kt:77]: First is "does this constructor have a `@Factory` annotation? If so, use it for constructing the object".
Second is "does _any other_ constructor have a `@Factory` annotation? If so, _do not_ use _this_ one".

Without that second case `@Factory` will conflict with the default constructor and an exception gets thrown.

File: keel-ec2-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ClusterSpec.kt (hunk 14 above)
Comment (robfletcher) [reply keel-ec2-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ClusterSpec.kt:111]: I just do what `gradle spotlessApply` tells me to

File: keel-titus-api/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/TitusClusterSpec.kt [hunk 15]
```diff
@@ -54,6 +56,45 @@ data class TitusClusterSpec(
...
+      env = env,
+      containerAttributes = containerAttributes,
+      iamProfile = iamProfile,
+      migrationPolicy = migrationPolicy,
+      resources = resources,
+      tags = tags
+    ),
+    overrides = overrides,
+    container = container
+  )
```
Comment (robfletcher) [inline keel-titus-api/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/TitusClusterSpec.kt:96]: `@Factory` constructor for `TitusClusterSpec` instead of requiring a custom deserializer

File: keel-titus-plugin/src/main/kotlin/com/netflix/spinnaker/keel/titus/jackson/KeelTitusApiModule.kt [hunk 16]
```diff
@@ -16,22 +11,12 @@ fun ObjectMapper.registerKeelTitusApiModule(): ObjectMapper = registerModule(Kee
...
   }
 }
 
-internal object KeelTitusApiDeserializers : Deserializers.Base() {
-  override fun findBeanDeserializer(type: JavaType, config: DeserializationConfig, beanDesc: BeanDescription): JsonDeserializer<*>? =
-    when (type.rawClass) {
-      TitusClusterSpec::class.java -> TitusClusterSpecDeserializer()
-      else -> null
-    }
-}
```
Comment (robfletcher) [inline keel-titus-plugin/src/main/kotlin/com/netflix/spinnaker/keel/titus/jackson/KeelTitusApiModule.kt:33]: No more custom deserialization in the Titus module

File: keel-titus-api/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/TitusClusterSpec.kt [hunk 17]
```diff
@@ -54,6 +56,45 @@ data class TitusClusterSpec(
...
+  @Factory
+  constructor(
+    moniker: Moniker,
+    deployWith: ClusterDeployStrategy = RedBlack(),
+    locations: SimpleLocations,
+    container: ContainerProvider,
+    capacity: Capacity?,
+    constraints: TitusServerGroup.Constraints?,
+    env: Map<String, String>?,
+    containerAttributes: Map<String, String>?,
```
Comment (luispollo) [inline keel-titus-api/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/TitusClusterSpec.kt:68]: Are we possibly missing `container` here again?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/KeelApiModule.kt (hunk 13 above)
Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/KeelApiModule.kt:77]: I'll add a comment to that effect

File: keel-titus-api/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/TitusClusterSpec.kt [hunk 18]
```diff
@@ -54,6 +56,45 @@ data class TitusClusterSpec(
...
+    deployWith = deployWith,
+    locations = locations,
+    _defaults = TitusServerGroupSpec(
+      capacity = capacity,
+      capacityGroup = capacityGroup,
+      constraints = constraints,
+      dependencies = dependencies,
+      entryPoint = entryPoint,
+      env = env,
+      containerAttributes = containerAttributes,
```
Comment (luispollo) [inline keel-titus-api/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/TitusClusterSpec.kt:88]: Are we possibly missing `container` here again?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/KeelApiModule.kt (hunk 13 above)
Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/KeelApiModule.kt:77]: I actually meant to pull this out as a proper class instead of an anonymous object too.

File: keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocCompatibilityTests.kt [hunk 19]
```diff
@@ -33,90 +29,55 @@ import strikt.api.expectThat
...
+//        "Titus cluster" to "/examples/titus-cluster-example.yml",
+//        "Titus cluster with artifact" to "/examples/titus-cluster-with-artifact-example.yml"
+//      ).forEach { (description, path) ->
+//        test("$description example is valid") {
+//          expectThat(validate(path))
+//            .describedAs(description)
+//            .isValid()
+//        }
+//      }
+//    }
```
Comment (luispollo) [inline keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocCompatibilityTests.kt:61]: Remove?

File: keel-titus-api/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/TitusClusterSpec.kt (hunk 18 above)
Comment (robfletcher) [reply keel-titus-api/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/TitusClusterSpec.kt:88]: No, look at line 95

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/FactoryAnnotationIntrospector.kt [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+class FactoryAnnotationIntrospector : NopAnnotationIntrospector() {
+  override fun findCreatorAnnotation(config: MapperConfig<*>, a: Annotated): JsonCreator.Mode? =
+    when {
+      // if this constructor has @Factory annotation, use it to create the object
+      a.hasAnnotation(Factory::class.java) -> PROPERTIES
+      // if *any other* constructor has @Factory annotation, ignore this one
+      a.rawType.constructors.any { it.isAnnotationPresent(Factory::class.java) } -> DISABLED
+      // in any other case fall back to the default
+      else -> super.findCreatorAnnotation(config, a)
+    }
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/FactoryAnnotationIntrospector.kt:23]: @luispollo I extracted the Factory handler to here and added comments

File: keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocCompatibilityTests.kt (hunk 11 above)
Comment (luispollo) [reply keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocCompatibilityTests.kt:36]: Is my understanding correct that a `SubmittedDeliveryConfig` successfully parsed with Jackson will have met all the rules in the schema (assuming we've configured Jackson to match the schema)? If not, it would be nice to add a filter that validates the JSON against the schema before it reaches `DeliveryConfigController`.

File: keel-schema-generator/src/test/kotlin/com/netflix/spinnaker/keel/schema/GeneratorTests.kt [hunk 21]
```diff
@@ -0,0 +1,950 @@
...
+
+  @Nested
+  @DisplayName("enum properties")
+  class EnumProperties : GeneratorTestBase() {
+    data class Foo(
+      val size: Size
+    )
+
+    enum class Size {
+      tall, grande, venti
```
Comment (luispollo) [inline keel-schema-generator/src/test/kotlin/com/netflix/spinnaker/keel/schema/GeneratorTests.kt:118]: ☕  😄

File: keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocCompatibilityTests.kt (hunk 11 above)
Comment (robfletcher) [reply keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocCompatibilityTests.kt:36]: Well, the schema is generated from the classes and parsing rules, but no we're not explicitly validating although we certainly could. Maybe a future change?

**Review by luispollo — APPROVED** [review]
I can't say that I reviewed the generator code, but the tests look very thorough (as usual!). Great job cleaning up all that custom code. ❤️

File: keel-ec2-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ClusterSpec.kt (hunk 14 above)
Comment (luispollo) [reply keel-ec2-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ClusterSpec.kt:111]: 🤷‍♂️

File: keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocCompatibilityTests.kt (hunk 11 above)
Comment (luispollo) [reply keel-web/src/test/kotlin/com/netflix/spinnaker/keel/apidocs/ApiDocCompatibilityTests.kt:36]: Sure, future PR sounds good. 👍

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/schema/Factory.kt (hunk 2 above)
Comment (luispollo) [reply keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/schema/Factory.kt:6]: Last try: `SchemaConstructor`? 🙂

**Comment by robfletcher** [comment]
> I can't say that I reviewed the generator code, but the tests look very thorough (as usual!). Great job cleaning up all that custom code. ❤️

My hope is that it's more comprehensible than the previous mechanism. If you feel like it needs some more commenting or breaking down of methods, I can do that.
